File: web_portal/backend/routes/verification.py
"""
Public certificate verification routes
"""

import os
import json
import sys
from pathlib import Path
from flask import Blueprint, request, jsonify, current_app
from werkzeug.utils import secure_filename

# Add certificate_utils to path
sys.path.append(str(Path(__file__).parent.parent.parent.parent))
from certificate_utils.verifier import CertificateVerifier, VerificationResult

verification_bp = Blueprint('verification', __name__)


def allowed_file(filename):
    """Check if file extension is allowed"""
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in current_app.config['ALLOWED_EXTENSIONS']


@verification_bp.route('/id/<cert_id>', methods=['GET'])
def verify_by_id(cert_id):
    """Verify certificate by ID"""
    try:
        # Get client info for logging
        ip_address = request.remote_addr
        user_agent = request.headers.get('User-Agent', 'Unknown')
        
        # Fetch certificate from database
        cert_record = current_app.supabase.get_certificate_by_id(cert_id)
        
        if not cert_record:
            # Log verification attempt
            current_app.supabase.insert_verification_log(
                cert_id, 'not_found', ip_address, user_agent
            )
            
            return jsonify({
                'status': 'NotFound',
                'message': 'No certificate found with this ID',
                'cert_id': cert_id
            }), 404
        
        # Check status
        if cert_record.get('status') == 'revoked':
            current_app.supabase.insert_verification_log(
                cert_id, 'revoked', ip_address, user_agent
            )
            
            return jsonify({
                'status': 'Revoked',
                'message': 'This certificate has been revoked',
                'cert_id': cert_id,
                'details': {
                    'device_name': cert_record.get('device_name'),
                    'wipe_date': cert_record.get('created_at')
                }
            })
        
        # Log successful verification
        current_app.supabase.insert_verification_log(
            cert_id, 'verified', ip_address, user_agent
        )
        
        return jsonify({
            'status': 'Verified',
            'message': 'Certificate is valid and authentic',
            'cert_id': cert_id,
            'details': {
                'device_name': cert_record.get('device_name'),
                'device_model': cert_record.get('device_model'),
                'wipe_method': cert_record.get('wipe_method'),
                'wipe_date': cert_record.get('wipe_start_time'),
                'status': cert_record.get('status')
            }
        })
        
    except Exception as e:
        current_app.logger.error(f"Verification error: {e}")
        return jsonify({'error': 'Verification failed'}), 500


@verification_bp.route('/file', methods=['POST'])
def verify_by_file():
    """Verify certificate by uploading JSON file"""
    try:
        # Check if file is in request
        if 'file' not in request.files:
            return jsonify({'error': 'No file provided'}), 400
        
        file = request.files['file']
        
        if file.filename == '':
            return jsonify({'error': 'No file selected'}), 400
        
        if not allowed_file(file.filename):
            return jsonify({'error': 'Invalid file type. Only JSON and PDF allowed'}), 400
        
        # Save file temporarily
        filename = secure_filename(file.filename)
        upload_path = Path(current_app.config['UPLOAD_FOLDER']) / filename
        file.save(str(upload_path))
        
        try:
            # Initialize verifier
            verifier = CertificateVerifier(
                public_key_path=current_app.config.get('PUBLIC_KEY_PATH')
            )
            
            # Verify file
            is_valid, message, cert_data = verifier.verify_certificate_file(str(upload_path))
            
            if not cert_data:
                return jsonify({
                    'status': 'Invalid',
                    'message': message
                }), 400
            
            # Get cert_id
            cert_id = cert_data.get('cert_id')
            
            # Check against database
            db_record = None
            db_match = False
            
            if cert_id:
                db_record = current_app.supabase.get_certificate_by_id(cert_id)
                if db_record:
                    db_is_valid, db_message = verifier.verify_against_database(cert_data, db_record)
                    db_match = db_is_valid
            
            # Log verification
            ip_address = request.remote_addr
            user_agent = request.headers.get('User-Agent', 'Unknown')
            result_status = 'verified' if is_valid and db_match else 'invalid'
            
            if cert_id:
                current_app.supabase.insert_verification_log(
                    cert_id, result_status, ip_address, user_agent
                )
            
            # Build result
            result = VerificationResult(
                is_valid=is_valid,
                message=message,
                cert_data=cert_data,
                db_match=db_match,
                details={
                    'signature_valid': is_valid,
                    'database_match': db_match,
                    'database_status': db_record.get('status') if db_record else None
                }
            )
            
            return jsonify(result.to_dict())
            
        finally:
            # Clean up uploaded file
            if upload_path.exists():
                upload_path.unlink()
        
    except Exception as e:
        current_app.logger.error(f"File verification error: {e}")
        return jsonify({'error': 'File verification failed'}), 500


@verification_bp.route('/hash', methods=['POST'])
def verify_by_hash():
    """Verify certificate by verification hash"""
    try:
        data = request.get_json()
        
        if not data or 'hash' not in data:
            return jsonify({'error': 'Hash is required'}), 400
        
        verification_hash = data['hash']
        
        # Search by verification hash
        response = current_app.supabase.client.table('certificates')\
            .select('*')\
            .eq('verification_hash', verification_hash)\
            .single()\
            .execute()
        
        if not response.data:
            return jsonify({
                'status': 'NotFound',
                'message': 'No certificate found with this hash'
            }), 404
        
        cert_record = response.data
        
        # Log verification
        ip_address = request.remote_addr
        user_agent = request.headers.get('User-Agent', 'Unknown')
        
        current_app.supabase.insert_verification_log(
            cert_record.get('cert_id'), 'verified', ip_address, user_agent
        )
        
        return jsonify({
            'status': 'Verified',
            'message': 'Certificate found and verified',
            'cert_id': cert_record.get('cert_id'),
            'details': {
                'device_name': cert_record.get('device_name'),
                'wipe_method': cert_record.get('wipe_method'),
                'wipe_date': cert_record.get('wipe_start_time')
            }
        })
        
    except Exception as e:
        current_app.logger.error(f"Hash verification error: {e}")
        return jsonify({'error': 'Verification failed'}), 500

File: web_portal/backend/routes/verification.js
// web_portal/backend/routes/verification.js
import express from 'express';
import multer from 'multer';
import { unlink } from 'fs/promises';
import { join } from 'path';
import supabaseClient from '../lib/supabaseClient.js';

const router = express.Router();

// Configure multer for file uploads
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const config = req.app.get('config');
    cb(null, config.uploadFolder);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
    cb(null, uniqueSuffix + '-' + file.originalname);
  }
});

const upload = multer({
  storage,
  fileFilter: (req, file, cb) => {
    const allowedTypes = ['application/pdf', 'application/json'];
    if (allowedTypes.includes(file.mimetype)) {
      cb(null, true);
    } else {
      cb(new Error('Invalid file type. Only PDF and JSON allowed'));
    }
  },
  limits: {
    fileSize: 16 * 1024 * 1024 // 16MB
  }
});

// Verify certificate by ID
router.get('/id/:certId', async (req, res) => {
  try {
    const { certId } = req.params;
    
    // Get client info for logging
    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.headers['user-agent'] || 'Unknown';

    // Fetch certificate from database
    const certRecord = await supabaseClient.getCertificateById(certId);

    if (!certRecord) {
      // Log verification attempt
      await supabaseClient.insertVerificationLog(
        certId,
        'not_found',
        ipAddress,
        userAgent
      );

      return res.status(404).json({
        status: 'NotFound',
        message: 'No certificate found with this ID',
        cert_id: certId
      });
    }

    // Check status
    if (certRecord.status === 'revoked') {
      await supabaseClient.insertVerificationLog(
        certId,
        'revoked',
        ipAddress,
        userAgent
      );

      return res.json({
        status: 'Revoked',
        message: 'This certificate has been revoked',
        cert_id: certId,
        details: {
          device_name: certRecord.device_name,
          wipe_date: certRecord.created_at
        }
      });
    }

    // Log successful verification
    await supabaseClient.insertVerificationLog(
      certId,
      'verified',
      ipAddress,
      userAgent
    );

    res.json({
      status: 'Verified',
      message: 'Certificate is valid and authentic',
      cert_id: certId,
      details: {
        device_name: certRecord.device_name,
        device_model: certRecord.device_model,
        wipe_method: certRecord.wipe_method,
        wipe_date: certRecord.wipe_start_time,
        status: certRecord.status
      }
    });
  } catch (error) {
    console.error('Verification error:', error);
    res.status(500).json({ error: 'Verification failed' });
  }
});

// Verify certificate by uploading file
router.post('/file', upload.single('file'), async (req, res) => {
  let uploadPath = null;

  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file provided' });
    }

    uploadPath = req.file.path;

    // Here you would implement the certificate verification logic
    // For now, we'll return a placeholder response
    // You would typically:
    // 1. Parse the file (PDF or JSON)
    // 2. Extract certificate data
    // 3. Verify signature
    // 4. Check against database

    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.headers['user-agent'] || 'Unknown';

    // Placeholder verification result
    const isValid = true;
    const message = 'Certificate verification not fully implemented';
    
    // For demonstration, assuming we extract cert_id from file
    const certId = 'demo-cert-id'; // This would come from parsing the file
    
    res.json({
      status: isValid ? 'Verified' : 'Invalid',
      message,
      details: {
        signature_valid: isValid,
        database_match: false,
        filename: req.file.originalname
      }
    });
  } catch (error) {
    console.error('File verification error:', error);
    res.status(500).json({ error: 'File verification failed' });
  } finally {
    // Clean up uploaded file
    if (uploadPath) {
      try {
        await unlink(uploadPath);
      } catch (err) {
        console.error('Failed to delete temporary file:', err);
      }
    }
  }
});

// Verify certificate by verification hash
router.post('/hash', async (req, res) => {
  try {
    const { hash } = req.body;

    if (!hash) {
      return res.status(400).json({ error: 'Hash is required' });
    }

    // Search by verification hash
    const { data, error } = await supabaseClient.client
      .from('certificates')
      .select('*')
      .eq('verification_hash', hash)
      .single();

    if (error || !data) {
      return res.status(404).json({
        status: 'NotFound',
        message: 'No certificate found with this hash'
      });
    }

    const certRecord = data;

    // Log verification
    const ipAddress = req.ip || req.connection.remoteAddress;
    const userAgent = req.headers['user-agent'] || 'Unknown';

    await supabaseClient.insertVerificationLog(
      certRecord.cert_id,
      'verified',
      ipAddress,
      userAgent
    );

    res.json({
      status: 'Verified',
      message: 'Certificate found and verified',
      cert_id: certRecord.cert_id,
      details: {
        device_name: certRecord.device_name,
        wipe_method: certRecord.wipe_method,
        wipe_date: certRecord.wipe_start_time
      }
    });
  } catch (error) {
    console.error('Hash verification error:', error);
    res.status(500).json({ error: 'Verification failed' });
  }
});

export default router;